test/shared/src/main/scala-3/zio/test/SpecLayerMacros.scala

```diff
@@ -27,10 +34,24 @@ object SpecLayerMacros {
   def provideSharedImpl[R0: Type, R: Type, E: Type](spec: Expr[Spec[R, E]], layer: Expr[Seq[ZLayer[_, E, _]]])(using
     Quotes
   ): Expr[Spec[R0, E]] = {
-    val expr = LayerMacros.constructStaticLayer[R0, R, E](layer)
+    import quotes.reflect._
+    val expr = LayerMacros.constructStaticProvideLayer[R0, R, E](layer)
     '{ $spec.provideLayerShared($expr) }
   }
 
+  def provideSomeSharedImpl[R0: Type, R: Type, E: Type](spec: Expr[Spec[R, E]], layer: Expr[Seq[ZLayer[_, E, _]]])(using
+    Quotes
+  ): Expr[Spec[R0, E]] = {
+    import quotes.reflect._
+    val expr: Expr[ZLayer[R0, E, ?]] = LayerMacros.constructStaticProvideSomeSharedLayer[R0, R, E](layer)
+    '{
+      // Contract of constructStaticProvideSomeSharedLayer ensures this
+      // I cannot obtain `?` from ZLayer, so I'm using `Any`
+      given <:<[R0 & Any, R] = null
```

Comment:
This worries me a bit, and I'm not entirely sure why it's needed or how does it work. `<:<` is not just an evidence class, it contains methods that are invoked implicitly in conversions. e.g., this method:

```scala
def foo[A](a: List[A])(using ev: <:<[A, Int]): List[Int] =
  a.map(_ + 2)
```

Decompiles to:

```java
    public <A> List<Object> foo(final List<A> a, final .less.colon.less<A, Object> ev) {
        return a.map((_$1) -> BoxesRunTime.unboxToInt(ev.apply(_$1)) + 2);
    }
```

As you can see, the `apply` method of `<:<` is being invoked. In the case where we use `null` that would cause a NPE, e.g.,:

```scala
object Foo {
  def foo[A](a: List[A])(using ev: <:<[A, Int]): List[Int] =
    a.map(_ + 2)

  @main def m = {
    given <:<[Any, Int] = null
    foo(List(1, 2, 3))
  }
}
```

Output:

```
Exception in thread "main" java.lang.NullPointerException: Cannot invoke "scala.$less$colon$less.apply(Object)" because "ev$1" is null
	at Foo$.foo$$anonfun$1(main.scala:7)
	at scala.collection.immutable.List.map(List.scala:247)
	at Foo$.foo(main.scala:7)
	at Foo$.m(main.scala:12)
	at m.main(main.scala:9)
```

Reply:
It's needed because provideSomeLayer requires evidence for type checking only. Internally it's doing it's own unsafe casting.

Later today maybe I'll describe why I need this ev in more detail.

Reply:
So the issue is with dynamic type that provideSome macros generate.

https://scastie.scala-lang.org/Dcfjq1ZOQ3CAhcBZ4hemdQ

We need 3 types:
- 1. Requirements of Spec
- 2. Layer type
- 3. Rest of requirenments

We do have 2 of them
1. Because Spec has it
3. Because user provides it in `[]` (with Auto this will also be auto generated)

But we cannot say ZLayer from LayerBuilder will have some type that which when we combine it (using `with`) with user provided type we will get type from Spec.

I still don't understand how Scala 2 does it - I believe it just ignores types altogether - I don't know if in Scala 3 I can make macro "typeless" and delay type resolution until actually running the macro

Reply:
Ok I see, thanks for the clarifications! I had a better look at the code and you're right, the class is only used as an evidence so there's no risk for an NPE.

One last comment, does this work? `given` produces a lot of byte code so inlining it is much more lightweight as it'll simply use `null` wherever it needs to be used.

```suggestion
      inline given <:<[R0 & Any, R] = null
```

Reply:
I'd like to get this PR merged in so I'll approve it as is and do the fix in a followup PR myself. I think there are a few places we could inline givens like this as well